### PR TITLE
KAFKA-2977: Transient Failure in kafka.log.LogCleanerIntegrationTest.cleanerTest

### PR DESCRIPTION
**Pull request by jinxing64**
Make MinCleanableDirtyRatioProp configurable(default 0.0F)in makeCleaner, thus log cleaning is always undergoing;
Also removed minDirtyMessages.

### CODE REVIEW

**Comment by guozhangwang** [comment]
Thanks @ZoneMayor, this makes sense to me. However in terms of the fix could we make MinCleanableDirtyRatioProp configurable in makeCleaner and set it to 0 in testCleaner so that it will always trigger cleaning regardless of the ratio?

File: core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala [hunk 1]
```diff
@@ -59,9 +59,10 @@ class LogCleanerIntegrationTest(compressionCodec: String) {
     val firstDirty = log.activeSegment.baseOffset
     // wait until we clean up to base_offset of active segment - minDirtyMessages
     cleaner.awaitCleaned("log", 0, firstDirty)
-
+    val compactedSize = log.logSegments.map(_.size).sum
     val lastCleaned = cleaner.cleanerManager.allCleanerCheckpoints.get(TopicAndPartition("log", 0)).get
-    assertTrue("log cleaner should have processed up to offset " + firstDirty, lastCleaned >= firstDirty);
+    assertTrue("log cleaner should have processed up to offset %d, but lastCleaned=%d".format(firstDirty, lastCleaned), lastCleaned >= firstDirty);
```
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala]: We can remove ";" here, and ditto below.

**Comment by jinxing64** [comment]
@guozhangwang  already removed the semicolons

**Comment by guozhangwang** [comment]
@ZoneMayor What do you think about the suggestion to change MinCleanableDirtyRatioProp in my last comment?

File: core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala [hunk 2]
```diff
@@ -59,9 +59,10 @@ class LogCleanerIntegrationTest(compressionCodec: String) {
     val firstDirty = log.activeSegment.baseOffset
     // wait until we clean up to base_offset of active segment - minDirtyMessages
     cleaner.awaitCleaned("log", 0, firstDirty)
-
+    val compactedSize = log.logSegments.map(_.size).sum
     val lastCleaned = cleaner.cleanerManager.allCleanerCheckpoints.get(TopicAndPartition("log", 0)).get
-    assertTrue("log cleaner should have processed up to offset " + firstDirty, lastCleaned >= firstDirty);
+    assertTrue("log cleaner should have processed up to offset %d, but lastCleaned=%d".format(firstDirty, lastCleaned), lastCleaned >= firstDirty)
+    assertTrue("log should have been compacted:  startSize=%d compactedSize=%d".format(startSize, compactedSize), startSize > compactedSize)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala]: Would you consider using String interpolation instead? For example:

```
assertTrue(s"log cleaner should have processed up to offset $firstDirty, but lastCleaned=$lastCleaned", lastCleaned >= firstDirty)
assertTrue(s"log should have been compacted: startSize=$startSize compactedSize=$compactedSize"
```

**Comment by jinxing64** [comment]
@ijuma 
using string interpolation now

File: core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala [hunk 3]
```diff
@@ -138,6 +139,8 @@ class LogCleanerIntegrationTest(compressionCodec: String) {
       logProps.put(LogConfig.SegmentIndexBytesProp, 100*1024: java.lang.Integer)
       logProps.put(LogConfig.FileDeleteDelayMsProp, deleteDelay: java.lang.Integer)
       logProps.put(LogConfig.CleanupPolicyProp, LogConfig.Compact)
+      logProps.put(LogConfig.MinCleanableDirtyRatioProp, 0F: java.lang.Float)
```
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/log/LogCleanerIntegrationTest.scala]: Actually can we make this configurable in `makeCleaner` (default to 0.0f) and replace `minDirtyMessages` which is not used any more? I think in an older version we use minDirtyMessages but changed it to a ratio, but never changed this function..

Also we can update the comment of "wait until we clean up to base_offset of active segment - minDirtyMessages" above.

**Comment by jinxing64** [comment]
@guozhangwang 
thanks a lot for feedback, really helpful;
I made MinCleanableDirtyRatioProp configurable in makeCleaner, also removed minDirtyMessages.

**Comment by guozhangwang** [comment]
@ZoneMayor Could you modify the first comment of this PR since it will be included in the commit message? I copied it here just for the record:

```
Code is as below:
val appends = writeDups(numKeys = 100, numDups = 3, log, CompressionCodec.getCompressionCodec(compressionCodec))
cleaner.startup()
val firstDirty = log.activeSegment.baseOffset
cleaner.awaitCleaned("log", 0, firstDirty)
....
val appends2 = appends ++ writeDups(numKeys = 100, numDups = 3, log, CompressionCodec.getCompressionCodec(compressionCodec))
val firstDirty2 = log.activeSegment.baseOffset
cleaner.awaitCleaned("log", 0, firstDirty2)

log cleaner and writeDups are two different threads;
log cleaner do cleaning every 15s, timeout in "cleaner.awaitCleaned" is 60s;
there is a filtering condition for a log to be chosen to become a cleaning target: cleanableRatio> 0.5(configured log.cleaner.min.cleanable.ratio) by default;
It may happen that, during "val appends2 = appends ++ writeDups(numKeys = 100, numDups = 3, log, CompressionCodec.getCompressionCodec(compressionCodec))", log is also undergoing a cleaning process; 
Since the segment size configured in this test is quite small: 100, there is possibility that before the end of 'writeDups', some 'dirty segment' of the log is already cleaned;
With tiny dirty part left, cleanableRatio> 0.5 cannot be satisfied;
thus firstDirty2>lastCleaned2, which leads to this test failed;
To fix, make MinCleanableDirtyRatioProp configurable in makeCleaner;
Also removed minDirtyMessages.
```

**Comment by jinxing64** [comment]
@guozhangwang 
yes, already modified;

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk and 0.9.0.

**Pull request closed without merging**